### ISSUE TITLE
SerialPIO receive at 9600bps shifts bits read

### ISSUE THREAD

**Issue by drakejest** [earlephilhower/arduino-pico#1276]
Im using SerialPIO for UART but when i print the response from the UART device serial monitor shows two questionmark characters "??". I know that its not supposed to be there because if i transfer it a hardware UART say Serial1 or Serial2 the question marks dont show up in the serial monitor

![image](https://user-images.githubusercontent.com/62618765/223099891-364c3082-639e-4d1b-8bd3-1c36c688575b.png)

The code:

```
SerialPIO dev1(8, 9, 512);

void setup() {

  pinMode(10, OUTPUT);
  digitalWrite(10, HIGH);
  
  dev1.begin(9600);
  Serial.begin(9600);

  while(!Serial);
  
  Serial.println("Device 3: ");
  dev1.println("AT");
  dev1.flush();
  delay(2000);

  Serial.println(dev1.readString());
}


void loop() {        
}
```

I went on investigating further by doing:

```
  //Serial.println(dev1.readString());
  while (dev1.available()) {
    Serial.println(dev1.read(),HEX);
  }
```
![image](https://user-images.githubusercontent.com/62618765/223101294-97b84916-eec0-4aee-8b26-893026883fcd.png)

moving over the device to a hardware uart its response is this:

![image](https://user-images.githubusercontent.com/62618765/223102457-c9133cfc-1ffb-4a37-90b6-518ba18f665b.png)

The last two characters are different, i checked with the ascii table what those ment its a non standard character, The hardware UARTs response is the correct one 0x0D 0x0A.

**Comment by earlephilhower** [comment]
The SerialPIO and the UART handle mis-framed bytes differently.  Can you grab a logic analyzer trace of your setup?  Without your hardware, it's impossible to reproduce this since you're having trouble with what some other chip is sending you.

What it looks like is you have a clock mismatch.  0x86 * 2 = 0x0d, 0x85 * 2 = 0x0a.  So the 9600bps speed (which is seldom exactly 9600 due to clock limitations on both sides) is off, with the external chip slower than the PIO expects.

**Comment by drakejest** [comment]
![image](https://user-images.githubusercontent.com/62618765/223222727-cdef8298-33c5-4c8d-9ef3-a7fc91f82320.png)

screen shot, sorry but i dont have a UART module. I will update the image after i figure out how to install the uart module of the analyzer

**Comment by drakejest** [comment]
![image](https://user-images.githubusercontent.com/62618765/223223611-8c639518-c30b-4f8b-bbcb-41958c2d082b.png)

Updated screenshot, and it confirms that PIO is recieving wrong data.

I have 6 UART device 2 hardware and 4 PIOs. all 4 PIOS has the last 2 characters wrong. If it is what you say a clock mismatch, what can be done?

**Comment by earlephilhower** [comment]
If you could get some measures of bit times that would help, both for PIO and for the external chip.  Then you could adjust the PIO baud up/down to better match the sender.

**Comment by drakejest** [comment]
for the PIO (sender)

![image](https://user-images.githubusercontent.com/62618765/223225541-2c08d5d7-b2e8-4932-9ef7-7e6343c3403d.png)

name	                type	start_time	duration	                data
Async Serial [1]	data	0.799333958	0.000989583333	0x41
Async Serial [1]	data	0.800480125	0.000989583333	0x54
Async Serial [1]	data	0.801626292	0.000989583333	0x0D
Async Serial [1]	data	0.802772458	0.000989583333	0x0A

For the device (response)

![image](https://user-images.githubusercontent.com/62618765/223225614-a07d2afc-68b2-4659-9292-86fe5c54b740.png)

name	        type	start_time	duration	                data
Async Serial	data	0.808673792	0.000989583333	0x41
Async Serial	data	0.809707792	0.000989583333	0x54
Async Serial	data	0.810741833	0.000989583333	0x0D
Async Serial	data	0.811775792	0.000989583333	0x0D
Async Serial	data	0.81280975	0.000989583333	0x0A
Async Serial	data	0.81384375	0.000989583333	0x4F
Async Serial	data	0.814877792	0.000989583333	0x4B
Async Serial	data	0.815911792	0.000989583333	0x0D
Async Serial	data	0.81694575	0.000989583333	0x0A




Im not sure if this is what you mean. I used an image to make the formatting easier. I also included a text form incase you want to copy and paste it somewhere

**Comment by drakejest** [comment]
Does the serial PIO accepts awkward baudrates like  9610 for example?

**Comment by earlephilhower** [comment]
> Does the serial PIO accepts awkward baudrates like 9610 for example?

Yes, it doesn't have any special speed requirements.

**Comment by drakejest** [comment]
okay got the correct data now at baud 9620, but at baud 9800 data is also correct. Should i go higher?  what is generally a good number here?

**Comment by drakejest** [comment]
Thank you for the help as always :)

**Comment by earlephilhower** [comment]
No worries.  The bugger is this should be an easy thing for the PIO to grab because technically you're to sample at the midpoint of a bit time.  So you have a **lot** of clock cycles to move fwd and back before you slip a bit like this.  Maybe we're grabbing at the end of a bit so it's hairy.  Worth investigating at some point...

**Comment by maxgerhardt** [comment]
Does the PIOSerial do like a 8-times oversampling and majority vote on the bit value?

**Comment by earlephilhower** [comment]
No, the PIO can't do anything so complicated.

Now, running at Fsys, the PIO waits for a 1->0 transition.  It then waits for 1/2 a bit time's worth of Fsys clks and captures 18(? from memory, may be 19) samples every 1/2 bit time and eventually pushes it to the core FIFO.  The ARM then decodes it and checks framing/parity/etc.

It should be possible to add a debug 1-bit GPIO sideset to indicate the sample clock to experimentally verify things.  What is weird here is that the PIO is being used at much higher frequencies by folks (and I tested at up to 1MHz IIRC).  At 9600bps there should be tons of margin.

It could also be a function of stop bit processing since the earlier chars seem to come through OK.  But I would have thought if that would cause the read data to shift left, not right like in this case.

**Comment by drakejest** [comment]
for the record, the device i am using is the SIM800L V2. Im using 6 of these connected to a single pico board. I have 4 serialPIO instance and all 4 of them wrongly reads the last 2 character.

**Comment by drakejest** [comment]
I bought new pico and pico W.  Its still does this, adjusting the baud rate solves the problem as proposed by @earlephilhower  :)

**Comment by earlephilhower** [comment]
@drakejest can you please reduce the BPS in your app back to 9600 exactly, and try the 1-character patch in #1315.  You can manually change the `34` to `35` in that one source line if you have trouble using the GIT version or managing PR pulls.

Essentially it looks like the PIO UART was capturing every 1/2 cycle, starting at the middle of the start bit.  But, that means that the first *2* bits captured (middle of start + end start/begin bit0) should be thrown out, not *1*.  Depending on hold times it could "just work", and speeding up the BPS (like you did to make it work) would also push that sample closer to the center of the bit and be more solid.

**Comment by drakejest** [comment]
Sorry took so long to reply, my modules are currently deployed. Will come back with an update when i get the chance to take it down. 

Thanks for fixing the problem :)

**Comment by drakejest** [comment]
@earlephilhower  so i tested out the new edit and there is a problem, all the characters that were correct is now wrong and the last two that were wrong is now correct 

![image](https://user-images.githubusercontent.com/62618765/230729117-1708e93a-a4c9-4420-9125-3c8b48df4f56.png)

reverting back to `decode >>= 33 - _rxBits;` goes back to the original problem

![image](https://user-images.githubusercontent.com/62618765/230729189-a82b06f1-25ff-4550-9017-e4470d7b14ba.png)

**Comment by earlephilhower** [comment]
Weird.  Loopback from another serial port on the pico showed no errors with this patch (or the unmodified one).  I'll close this and the PR since there's something special about that chip...

**Comment by Apehaenger** [comment]
Had the same issue with an [WT901 IMU](https://witmotion-sensor.com/products/9-axis-accelerometer-tilt-sensor-wt901-high-accuracy-acceleration-gyroscope-angle-magnetometer-with-kalman-filtering-triaxial-mpu9250-ahrs-imu-iic-ttl-200hz-for-pc-android-arduino).
Everything worked great with HardwareSerial but when trying to drive it via SerialPIO I got a couple of non-sense data. After I tried everything I googled around and (luckily) found this post. Couldn't believe at first, but tried with 9650 baud and got immediately correct data. Tried up to 9800 baud and all worked, but not 9600!
